core/src/main/scala/akka/persistence/jdbc/state/scaladsl/JdbcDurableStateStore.scala

```diff
@@ -86,56 +87,90 @@ class JdbcDurableStateStore[A](
   def deleteObject(persistenceId: String): Future[Done] =
     db.run(queries._delete(persistenceId).map(_ => Done))
 
-  def currentChanges(tag: String, offset: Offset): Source[DurableStateChange[A], NotUsed] = offset match {
-    case NoOffset      => makeSourceFromStateQuery(tag, None)
-    case APSequence(l) => makeSourceFromStateQuery(tag, Some(l))
-    case _             => ??? // should not reach here
+  def currentChanges(tag: String, offset: Offset): Source[DurableStateChange[A], NotUsed] =
+    currentChangesByTag(tag, offset.value)
+
+  def changes(tag: String, offset: Offset): Source[DurableStateChange[A], NotUsed] =
+    _changesByTag(tag, offset.value, terminateAfterOffset = None)
+
+  private def currentChangesByTag(tag: String, offset: Long): Source[DurableStateChange[A], NotUsed] = {
+    Source
+      .futureSource(maxJournalSequence().map { maxOrderingInDb =>
+        _changesByTag(tag, offset, terminateAfterOffset = Some(maxOrderingInDb))
+      })
+      .mapMaterializedValue(_ => NotUsed)
+  }
+
+  private def _currentChangesByTag(
+      tag: String,
+      from: Long,
+      batchSize: Long,
+      queryUntil: MaxOffset): Source[DurableStateChange[A], NotUsed] = {
+    if (queryUntil.value < from.value) Source.empty
+    else changesByTag(tag, from, queryUntil.value, batchSize).mapAsync(1)(Future.fromTry)
+  }
+
+  private def changesByTag(
```

Comment:
A bit confusing with one private method called `changesByTag` and one `_changesByTag`, giving them names that indicates how they are different would be nice.

Reply:
changed the names ..